src/wallet/sqlite.h

```diff
@@ -63,7 +64,16 @@ class SQLiteDatabase : public WalletDatabase
 
     const std::string m_file_path;
 
-    void Cleanup() noexcept;
+    /**
+     * This mutex protects SQLite initialization and shutdown.
+     * sqlite3_config() and sqlite3_shutdown() are not thread-safe (sqlite3_initialize() is).
+     * Concurrent threads that execute SQLiteDatabase::SQLiteDatabase() should have just one
+     * of them do the init and the rest wait for it to complete before all can proceed.
+     */
+    static Mutex g_sqlite_mutex;
+    static int g_sqlite_count GUARDED_BY(g_sqlite_mutex);
+
+    void Cleanup() noexcept EXCLUSIVE_LOCKS_REQUIRED(!g_sqlite_mutex);
```

Comment:
Not sure about this annotation as @MarcoFalke [pointed](https://github.com/bitcoin/bitcoin/pull/25491#issuecomment-1169990668):
> Annotations are disabled for con-/de-structors

Nevertheless, it would be nice to add `AssertLockNotHeld(g_sqlite_mutex);` to the function implemetation.

Reply:
Added `AssertLockNotHeld()`.

I kept the annotation on the `Close()` method, even though it is called only from the constructor and the destructor in case it gets called from elsewhere in the future.

This looks a bit strange:

```cpp
Mutex m1;
int global_counter GUARDED_BY(m1){0};

void f() EXCLUSIVE_LOCKS_REQUIRED(!m1)
{
    AssertLockNotHeld(m1);
    LOCK(m1);
    ++global_counter;
}

SQLiteDatabase::SQLiteDatabase()
{
    LOCK(m1);
    // ok!?
    f();

void SQLiteDatabase::Open() // any other method
{
    LOCK(m1);
    // error: cannot call function 'f' while mutex 'm1' is held [-Werror,-Wthread-safety-analysis]
    f();
```

I understand annotations on the constructor are disabled, but the above is more than that. Are constructors treated like `NO_THREAD_SAFETY_ANALYSIS`?

Reply:
> Are constructors treated like NO_THREAD_SAFETY_ANALYSIS?

Yes, see https://clang.llvm.org/docs/ThreadSafetyAnalysis.html#no-checking-inside-constructors-and-destructors